### ISSUE TITLE
Live map is missing offline stored caches

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#9374]
As verified by @rsudev there is still a bug in the synchronization of the caches in an area and the layerlist in the new map. It is not really reproducible, but it is always happening for the same caches that are stored offline and disappear while zooming...

Also see https://github.com/cgeo/cgeo/issues/8682#issuecomment-723238788 and following

**Comment by eddiemuc** [comment]
Hi @rsudev , @fm-sys 

from looking in the code (class `cgeo.geocaching.maps.mapsforge.v6.caches.StoredCachesOverlay`) I found that caches are only ever loaded from the CacheCache (method `DataStore.loadCachedInViewport` is used) and that the CacheCache size is limited to 1000 entries max.

So I wonder if the following sequence of events could trigger this bug:

1. LiveMap is opened, loading of stored caches and live caches is triggered simultaniously
2. Loading of stored caches finishes first, they are displayed
3. Loading of live caches finishes second. In the background, this function did load more than (or close to) 1000 caches (it loads caches for 3 times the size of the viewport from different providers!), such that some of the stored caches currently displayed are pushed out of the CacheCache
4. User is doing a rezoom. Again (re)loading of stored caches and live caches is triggered simultaniously
5. Again loading of stored caches finishes first -> but does not find all stored caches in CacheCache since they were pushed out earlier

Do you think that's a possible explanation?

**Comment by rsudev** [comment]
Might play in, but the scenarios I have in mind for sure only involve a substantially smaller number of caches (below 100).

**Comment by moving-bits** [comment]
> 3. Loading of live caches finishes second. In the background, this function did load more than (or close to) 1000 caches (it loads caches for 3 times the size of the viewport from different providers!)

At least for the gc.com connector, max caches returned by a query is 500, don't have the other connectors' limits at hand.

**Comment by eddiemuc** [comment]
> Might play in, but the scenarios I have in mind for sure only involve a substantially smaller number of caches (below 100).

Ok, so my thought was obviously wrong.

Still I find it strange that the live map only displays caches from CacheCache and not from the database.

**Comment by moving-bits** [comment]
> Still I find it strange that the live map only displays caches from CacheCache and not from the database.

AFAIK: `LiveCachesOverlay` is only for caches gathered live from the servers. Most of them are probably not yet stored in the database. Additionally there is a `StoredCachesOverlay` presenting those caches which are already stored in the database. (At least that's what I think to remember from the `NewMap` implementation, but @rsudev might be able to shed more light on this.)

**Comment by eddiemuc** [comment]
@moving-bits I was talking about `StoredCachesOverlay` the whole time. There, only caches from CacheCache are loaded, not from DB (at least if my code analysis is correct)

**Comment by moving-bits** [comment]
> @moving-bits I was talking about `StoredCachesOverlay` the whole time.

My bad, haven't followed this thread close enough.

> There, only caches from CacheCache are loaded, not from DB (at least if my code analysis is correct)

If I understand the code correctly, we have the following sequence:
- `StoredCachesOverlay:load()` calls 
- `DataStore.loadCachedInViewport()` which calls
- `DataStore.loadInViewport()` with parameter `stored=false`

According to the description this means to not only use caches stored offline, but _additionally_ use the CacheCache. (So maybe the parameter should be called "storedOnly".)

`@param stored {@code true} to query caches stored in the database, {@code false} to also use the CacheCache`

A query against the database is run in either cases, but with `stored=false` geocodes from the CacheCache are added to the result set.

**Comment by rsudev** [comment]
This was a design choice. Loading from database was (is?) quite slow in comparison, so the CacheCache is a buffer between the map (and other code parts as well) and the db. The size limit is a problem (was introduced in the light of average hardware specs some years ago). Addressing this topic would be a major advantage, but unfortunately is it problematic to get the suggested rooms-based implementation up to speed (understandable, I have the experience of trying to keep a larger overhaul over several month in sync with the quickly moving code base).
(@moving-bits was a bit faster with his answer)

**Comment by eddiemuc** [comment]
You're right, I misread code in `DataStore.loadInViewport()`. Sorry for that

**Comment by Lineflyer** [comment]
We had a user on support mail with similar problem:
- Imported a pocket query
- Used live map while moving, using offline map
- After a while waypoints of stored caches are still shown, however the stored caches theirselves are missing until he manually moves the map

**Comment by wolverine007** [comment]
I am facing a similar problem inside 2020.12.22-RC. 
During the last days I was also seeing one 161m distance circle around no cache icon while using the Google maps Sat layer, switching back to offline OSM, the icon was shown correctly again.

It seems that the app was loading the cache listing again, cause first the icon was shown with the correct online coordinates - so approximately 60m away from the offline stored ones. Then a refresh happend and the icon moved to the incorrect location stored offline, but than the icon suddenly disappeared.

Meanwhile I updated the offline version of this buggy cache and the issue is gone.

So maybe this issue is taking place in correlation to a difference between offline and online cache coords?

**Comment by Lineflyer** [comment]
One other report from a beta user (2020.12.22-RC) as well:
After some zoom in and out of the live map, some cache icons might be missing. I noticed it, because the red direction line (which I had activated before is showing but the cache at its endpoint was no longer been shown. I had to zoom in / move the map a bit to make it reappear. Not reproducible so far.

**Comment by Lineflyer** [comment]
Attached a screenshot from the user mentioned in my previous posting, which makes the situation very clear:
![Screenshot_2021-01-05-06-31-56-171_cgeo geocaching](https://user-images.githubusercontent.com/949669/103627447-eba52480-4f3d-11eb-951e-b6d9e257c604.jpg)

**Comment by fm-sys** [comment]
Ok, seams like OSM as well as GM is affected

**Comment by alan666notb** [comment]
I was out caching today - haven't been out for a while because of Covid - and I noticed this problem and it was really quite bad.
Almost every time I opened the map, I had to zoom out/in to make all the icons appear.
I'm on 2020.12.22-RC, probably the first time I've used this version out in the field.

**Comment by Lineflyer** [comment]
More than 10 support mails already a few hours after releasing 2021.01.10 regarding this issue.
Users state, that moving, zooming makes caches disappear and reappear randomly.

**Comment by Lineflyer** [comment]
I tried to find a reproducible scenario but failed so far.
However I managed to trigger the problem by just repeatedly zooming in/out and moving the map carefully back and forth.

The problem I have seen is, that 
- all saved caches vanish from the map
- all found (unsaved) caches vanish from the map

The log shows:
```
01-11 13:41:23.891 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC5HQ40
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC8YTEY
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6TPD4
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC8BMCK
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC7C6K8
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC7Q8TF
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC5HQ2Z
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC7MEY3
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC352Y3
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3EQ4W
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC85Z5Y
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC8P68F
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC7709Z
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC926AT
01-11 13:41:23.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC5HQ47
01-11 13:41:24.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:24.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:24.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:24.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:26.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:26.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:26.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:26.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:27.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:27.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:27.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:27.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:28.643 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:28.643 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:28.643 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:28.643 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:29.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:29.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:29.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:29.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:31.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:31.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:31.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:31.142 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:32.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:32.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:32.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:32.392 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:33.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:33.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:33.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:33.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:35.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:35.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:35.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:35.642 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29
01-11 13:41:36.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC150HM
01-11 13:41:36.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC3KGEJ
01-11 13:41:36.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC9304M
01-11 13:41:36.892 29880 19388 W cgeo    : [pool-6-thread-1] CacheCache.getInViewport: got cache with null coordinates: GC6JY29```

Without having checked in details I assume those GC-codes are the caches vanishing from the map.
The coordinates should not be NULL, because either those caches were saved locally or been retrieved in that session from the website live map.

**Comment by Lineflyer** [comment]
On a second test I could again trigger the problem. This time I do not see this NULL coordinates warning, but the following log shows, that when just zooming out and in and thus always using cached data (staying in the same (bigger) viewport), the resulting caches according to the log are fluctuating. Maybe that helps debugging the problem.

```
01-11 13:49:26.094 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#249):END (3ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#249[GC517B6,GC6V7YP,GC8VZG6,GCKK5X,GC7EEKM,GC36T38,GC735PV,GC73AY0,GC6RQ7Y,GC4ATH9](0ms);RET:#1(3ms);
01-11 13:49:26.098 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:26.099 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=249, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:26.511 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:27.342 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#25):START
01-11 13:49:27.348 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#25):END (6ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#25[GC5PXQJ,GC8YTEY,GC40D80,GC6TPD4,GC5HQ4B,GC8FDC4,GC7Q8TF,GC735T4,GC75RZW,GC5HQ47](0ms);RET:#1(6ms);
01-11 13:49:27.348 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:27.349 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=249, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:27.641 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:28.122 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:28.592 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):START
01-11 13:49:28.596 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):END (4ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#20[GC1ZXEZ,GC5HQ40,GC276PG,GC8YTEY,GC40D80,GC6TPD4,GC5HQ4B,GC7C6K8,GC7Q8TF,GC5HQ2Z](0ms);RET:#1(4ms);
01-11 13:49:28.597 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:28.597 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=249, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:28.642 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:29.591 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#8):START
01-11 13:49:29.594 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#8):END (3ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#8[GC8YTEY,GC40D80,GC7MEY3,GC352Y3,GC8P68F,GC926AT,GC7C6K8,GC7Q8TF](0ms);RET:#1(3ms);
01-11 13:49:29.597 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:29.597 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=8, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:29.601 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:30.067 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:30.564 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:30.841 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#58):START
01-11 13:49:30.844 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#58):END (3ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#58[GC52EBF,GC8YTEY,GC8H9C8,GC6TPD4,GC7EEKM,GC5HQ4B,GC8FDC4,GC5YREE,GC81KJ5,GC8RQ4P](0ms);RET:#1(3ms);
01-11 13:49:30.850 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:30.850 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=58, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:31.296 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:31.894 20250 20250 I chatty  : uid=10259(cgeo.geocaching) identical 1 line
01-11 13:49:32.575 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:32.592 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):START
01-11 13:49:32.598 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):END (6ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](1ms);gc#20[GC1ZXEZ,GC5HQ40,GC276PG,GC8YTEY,GC40D80,GC6TPD4,GC5HQ4B,GC7C6K8,GC7Q8TF,GC5HQ2Z](1ms);RET:#1(6ms);
01-11 13:49:32.599 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:32.599 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=58, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:33.289 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:33.842 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):START
01-11 13:49:33.844 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#20):END (3ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](1ms);gc#20[GC1ZXEZ,GC5HQ40,GC276PG,GC8YTEY,GC40D80,GC6TPD4,GC5HQ4B,GC7C6K8,GC7Q8TF,GC5HQ2Z](1ms);RET:#1(3ms);
01-11 13:49:33.844 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:33.845 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=20, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:33.887 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:34.518 20250 20250 D cgeo    : [main] bearing=0.0, tilt=0.0, canDisable=false
01-11 13:49:35.091 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#58):START
01-11 13:49:35.095 20250 20360 D cgeo    : [pool-6-thread-1] DataStore.loadCachesFromGeoCodes(#58):END (4ms)flags:[CACHE_AFTER, DB_MINIMAL, WAYPOINTS, OFFLINE_LOG](0ms);gc#58[GC52EBF,GC8YTEY,GC8H9C8,GC6TPD4,GC7EEKM,GC5HQ4B,GC8FDC4,GC5YREE,GC81KJ5,GC8RQ4P](0ms);RET:#1(3ms);
01-11 13:49:35.095 20250 20362 D cgeo    : [pool-5-thread-1] Filtering out 0 caches: []
01-11 13:49:35.096 20250 20362 D cgeo    : [pool-5-thread-1] searchByViewport: cached=true, results=20, viewport=(N 51Â° 16.838' Â· E 006Â° 57.315',N 51Â° 24.860' Â· E 007Â° 08.090')
01-11 13:49:37.128 20250 20276 D cgeo    : [looper callbacks] RotationProvider: unregistering listener
```

**Comment by Lineflyer** [comment]
@cgeo/team 
I would like to draw your attention to this one.
While for me this bug only shows up rarely, it seems we have a pretty big regression here compared to 2020.12.03 release looking at the amount of support mails.
Today the support inbox is flooded with mails regarding this problem. :(

Maybe one could dive into the relevant commits between 2020.12.03 and current release 2021.01.10?

**Comment by eddiemuc** [comment]
I tried already yesterday to get a handle on this one but were unable to reproduce it.

**Comment by Lineflyer** [comment]
Is this file also involved when drawing caches on the map?
https://github.com/cgeo/cgeo/blob/release/main/src/cgeo/geocaching/maps/google/v2/GoogleMapObjectsQueue.java

**Comment by Lineflyer** [comment]
Comments from an affected user (untested):

---
- it has nothing to do with a cache being stored or not. 
- the problem only if you set the switch from large to small icons to auto.  (Comment Lineflyer: or way more often)
- the problem only occurs with both Google maps, not with osm. 
---

**Comment by moving-bits** [comment]
> Definetely easier to trigger this bug on GMaps than on OSM (not even sure if it show up in OSM)

This bug occurs also on OSM. Have seen this "out in the wild" during geocaching, but have not been able to reproduce it in the emulator so far.

> What was changed in that area after 2020.12.03 ?

I've seen it already before the current release, so the change that triggered it must be older than 2020-12-03.
I'm not sure if it has been there in an even earlier release.

> I remember the accuracy circles code?!

That has been merged around Dec 20, so should not be the culprit.

**Comment by geocachermgo** [comment]
> Comments from an affected user (untested):
> 
>     * it has nothing to do with a cache being stored or not.
> 
>     * the problem only if you set the switch from large to small icons to auto.  (Comment Lineflyer: or way more often)
> 
>     * the problem only occurs with both Google maps, not with osm.

Connected with #9027 ?

**Comment by alan666notb** [comment]
Some info maybe of use,

- I'm sitting at home using OSM offline maps and with stored caches only.
- I can easily reproduce the problem by scrolling the map slowly, in small increments at a time
- The 'count' of caches actually on display is taken into account when deciding if to show additional waypoints. I.E. when there are many missing caches on display, extra waypoints like parking area and trailheads are shown. If I make a 'sudden scroll', the missing caches appear and the additional waypoint all disappear.

**Comment by moving-bits** [comment]
We may have two separate problems here:

#9565 indeed seems to trigger a refresh problem on GM. I'm able to reproduce this with the following sequence:
- open GM, set "compact icons" to "auto"
- zoom out until compact icons are shown
- rapidly zoom in - at some point in time no or only a few caches are shown, although many more should be visible
- if it doesn't work on first run, zoom out and back in a few times

To mitigate this, we'll need to revert #9565 (I'll prepare a PR for this, based to `release`) and find a different solution for #9552.

But as #9565 is only related to GM, and we definitively have a problem on OSM as well, although quite less frequently, there has to be a separate cause for that. I'm not able yet to reliably reproduce it on OSM, though, it just happened a few times while being out geocaching. Other than the error in part 1 the look is a bit different, it's not caches anywhere on the screen missing, but a sharp vertical line beside which caches are missing. In my case, it's been offline stored caches only, offline OSM map.

**Comment by Lineflyer** [comment]
> But as #9565 is only related to GM, and we definitively have a problem on OSM as well, although quite less frequently, there has to be a separate cause for that.

**Comment by moving-bits** [comment]
Thanks @alan666notb,

> Some info maybe of use,
> * I'm sitting at home using OSM offline maps and with stored caches only.
> * I can easily reproduce the problem by scrolling the map slowly, in small increments at a time
> * The 'count' of caches actually on display is taken into account when deciding if to show additional waypoints. I.E. when there are many missing caches on display, extra waypoints like parking area and trailheads are shown. If I make a 'sudden scroll', the missing caches appear and the additional waypoint all disappear.

very good description. Following it I'm able to reproduce on a physical device using OSM offline. 🚀 
Will try to analyze in the debugger the next days.

**Comment by moving-bits** [comment]
I think I found the reason for the behavior @alan666notb describes:

There is some code in `NewMap` code in `StoredCachesOverlay` that remembers the last viewport to detect movements. If the movement is large enough, a reload of caches for the new viewport is triggered.

Due to some change introduced (by me, I have to admit) with the automatic switching of compact icons (#8389) this code was altered slightly, which lead to the previous viewport being reset on every cycle, leading to slow moves being ignored instead of accumulated.
Something similar had been introduced with the same PR for the `LiveCachesOverlay`, but this has been corrected a while ago already. 

I will prepare a PR, which fixes that for the `StoredCachesOverlay` as well. With this line reverted, small movements do work, and automatic switching for compact icons works as well in my smoke test.

This is `NewMap` only, so after merging this we will need to retest, if there is still some misbehavior on either Google Maps or OSM.
